tools/clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5097,14 +5097,65 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
   }
 }
 
+void Sema::ValidateShaderAttribtues(Scope *S, Decl *D, const AttributeList *A) {
+  bool hasComputeAttr = false;
+  bool hasNodeAttr = false;
+  int attrNum = 0;
+
+  for (const AttributeList *l = A; l; l = l->getNext()) {
+    if (l->getKind() == AttributeList::AT_HLSLShader) {
+      Expr *E = l->getArgAsExpr(0);
+      if (E->isTypeDependent() || E->isValueDependent() ||
+          E->getStmtClass() != Stmt::StringLiteralClass) {
+        break;
+      }
+
+      StringLiteral *sl = cast<StringLiteral>(E);
+      std::string str = sl->getString();
+      if (str == "compute") {
+        hasComputeAttr = true;
+        attrNum++;
+      } else if (str == "node") {
+        hasNodeAttr = true;
+        attrNum++;
+      } else {
+        for (auto shaderAttrStr :
+             {"vertex", "pixel", "hull", "domain", "geometry", "raygeneration",
+              "intersection", "anyhit", "closesthit", "miss", "callable",
+              "mesh", "amplification"}) {
+          if (str == shaderAttrStr) {
+            attrNum++;
+            break;
```

Comment:
I think it would be better to keep a note of the unique attribute that you've already seen and emit an error if you find another one here that conflicts. This woudl avoid having to perform any additional checks. 

So if you had a `lastAttr` string initialized to empty, then when you found one of these, you could:
* assign it if the string is empty
* skip if it's identical
* skip if one is node and the other is compute
* error in any other case.